PayU Latam: Add new gateway

shasum opened the pull request:
@duff only purchase and authorize works in sandbox

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,414 @@
+        "diners_club" => "DINERS"
+      }
+
+      # Override defaults using options
```
I'm thinking we can remove these comments.

duff commented:
@shasum I'd love to look at this PR with all of the commented out tests removed.  It'll simplify things a bunch.  If support for those operations are ever added to the gateway sandbox, we can add tests then.  For now, a small comment in the remote test file explaining that only some operations are supported in the sandbox should suffice.

shasum commented:
@duff feedback updates and more tests that cover all supported Latam countries

duff commented:
I haven't forgotten about this one!

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,393 @@
+      # hence 100.00 instead of typical 1.00
+      def verify(credit_card, options={})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { purchase(10000, credit_card, options) }
```
I'm thinking we should go with `authorize` here even if some countries don't support it.  We wouldn't want a `verify` to actually charge someone's card since  the `void` could fail.  If a country doesn't support `authorize`, then it doesn't support `verify`.

duff replied:
I also wonder if $100 is the lowest in all countries.  That's a pretty high minimum!

shasum replied:
Agreed. There is a getting charged risk if/when `void` fails. Will switch to `authorize`. I wasn't sure if it was okay to implement an op that worked selectively. Now, I figure we just let users know about upstream gateway's limits in terms how things work.

duff commented on test/remote/gateways/remote_payu_latam_test.rb:
```diff
@@ -0,0 +1,213 @@
+                       phone: "7563126")
+    }
+
+    @options_brazil = {
```
I'm mulling on whether we need tests for each country.  Is there any coding difference in the gateway for each country, or is it mostly a different billing address and `account_id`?

shasum replied:
As there was lot of back n forth while testing this gateway, I had at various times tried out each different `account_id` that maps to a different country as you rightly guessed. In the final commit, I just left them all in.

Pattern wise, there is Brazil and the rest, so this can be trimmed down to just those two.

duff commented on test/remote/gateways/remote_payu_latam_test.rb:
```diff
@@ -0,0 +1,213 @@
+    assert response.test?
+  end
+
+  def test_successful_purchase_sans_options
```
This is super useful methinks!  👍

duff commented:
It would be great if we could also support `verify_credentials` in new gateways.  Examples:

https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/stripe.rb#L240

https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/authorize_net.rb#L164

https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/blue_snap.rb#L117

https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/braintree_blue.rb#L178

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,393 @@
+      def add_order(post, options)
+        order = {}
+        order[:accountId] = @options[:account_id]
+        order[:referenceCode] = options[:order_id] || SecureRandom.hex(16)
```
Could use this : 
https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateway.rb#L107

duff replied:
And in the other places where we're referring to `SecureRandom.hex(16)`.

shasum replied:
👍  had noticed it in [test_helper.rb](https://github.com/activemerchant/active_merchant/blob/master/test/test_helper.rb#L235), didn't realize there's a similar one under [gateway.rb](https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateway.rb#L107)

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,393 @@
+      end
+
+      def add_payer(post, options)
+        if address = options[:billing_address] || options[:address]
```
Thinking we can just use `options[:billing_address]` and not worry about `options[:address]`.

shasum replied:
`shipping_address` else `billing_address` else `address`

think I saw this address fallback sequence on another gateway and borrowed it.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,393 @@
+        end
+        buyer[:cnpj] = options[:cnpj] if options[:cnpj]
+        buyer[:emailAddress] = options[:shipping_email] || options[:email] || 'unspecified@example.com'
+        if address = options[:shipping_address] || options[:billing_address] || options[:address]
```
Thinking we can just rely on `options[:shipping_address]` here and disregard the other 2.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,393 @@
+          add_credentials(post, 'SUBMIT_TRANSACTION')
+          add_transaction_type(post, 'VOID')
+          add_reference(post, authorization)
+        when 'purchase', 'capture'
```
Usually `void` happens on authorizations and `refund` happens on captures and purchases.  I'm wondering if we can have a `def refund` and split it out into 2 operations.

shasum replied:
👍 

Cool. The association between `void-authorization` and `refund-purchase` makes sense. Noticing the gateway does not support partial refunds, I somehow bunched those two together.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,393 @@
+        if payment_method.is_a?(String)
+          _, brand, cvv, token = split_authorization(payment_method)
+          credit_card = {}
+          credit_card[:securityCode] = cvv
```
Can this work without the `cvv`?  Since the `cvv` can't be stored for PCI reasons, we can't keep it in the authorization.

duff replied:
Guess like a bunch of operations, it's difficult to tell since the sandbox doesn't support store.  😭

shasum replied:
Right. Can't think of a way to find out for sure without running it in prod. Gateway [docs](http://developers.payulatam.com/en/api/tokenization.html) has this to say about stored tokens:
> Note: It is impossible to process transactions without validation of the cvv2 security code for the following countries. You need to send the SecurityCode parameter in the request. Argentina, Mexico, Panama, Peru.

Brazil and Colombia seem to be excluded from this `cvv` requirement as per [docs](http://developers.payulatam.com/en/api/tokenization.html).

Do you see a clean alternative to NOT keeping `cvv` in the returned authorization and still supporting `store` for all countries (or) we choose to support `store` for only Brazil and Colombia for now?

duff replied:
I'm thinking our only option is to not store the cvv in the authorization.  This means it may work in some areas and not in others.

shasum replied:
😿  will go ahead and change

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,393 @@
+      def purchase(amount, payment_method, options={})
+        post = {}
+
+        add_credentials(post, 'SUBMIT_TRANSACTION')
```
Looks like there's some logic that shows up in `purchase` and `authorize`.  Perhaps it can be refactored a bit to DRY things up some.

shasum replied:
Yup, agreed, definitely needs some DRYing 👍

duff commented:
@shasum Getting closer!  👍 

Made a few comments / questions of things I noticed.

shasum commented:
@duff updated commit pushed

Previous feedback items have been addressed except for the following,
- Brazil fails when `shipping_address` is not provided. Hence, `billing_address` has been retained as a shipping default with override.
- Removing `cvv` from authorization. Refer previous comment thread to finalize this part.

duff commented:
So if a customer specifies no billing_address and no shipping address, then it'll work, right?

The Brazil shipping address thing is an odd one!  Does it happen if the country is specified for the billing address and Brazil is specified?  I don't think I've ever heard of a shipping_address being required.  What if nothing is being shipped?

I'm wondering if we can bag on using the billing_address as a default shipping address.  Then if a customer gets an error message about shipping address being required in some circumstances, then they might be able to adjust their gateway settings in their gateway UI to turn off such a requirement.  Or they can start passing shipping address in those circumstances.  I'm a bit cautious about an implicit default shipping address.  Whatcha think?

shasum commented:
>So if a customer specifies no billing_address and no shipping address, then it'll work, right?

For countries other than Brazil, it works, no address no problem. But, in case of Brazil, it fails Antifraud check if either address is omitted ( both addresses needed for successful Brazil txn )

>The Brazil shipping address thing is an odd one! Does it happen if the country is specified for the billing address and Brazil is specified? I don't think I've ever heard of a shipping_address being required. What if nothing is being shipped?

It happens when the Brazilian `account_id` is used along with Brazilian currency and billing address inputs. Under these circumstances it expects a non-null shipping address, else returns `ANTIFRAUD_REJECTED` error. Good question about nothing being shipped, not sure what the gateway expects in such cases.

>I'm wondering if we can bag on using the billing_address as a default shipping address. Then if a customer gets an error message about shipping address being required in some circumstances, then they might be able to adjust their gateway settings in their gateway UI to turn off such a requirement. Or they can start passing shipping address in those circumstances. I'm a bit cautious about an implicit default shipping address. Whatcha think?

Initially, I was hesitant as well in having such implicit defaults. Later, we had use cases like [this](https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/cyber_source.rb#L252-L262), which had me think it might be okay to go with such address related defaults.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,400 @@
+          buyer[:dniNumber] = dni_number
+        end
+        buyer[:cnpj] = options[:cnpj] if options[:cnpj]
+        buyer[:emailAddress] = options[:shipping_email] || options[:email] || 'unspecified@example.com'
```
I wonder if we can simplify this to just:

```
buyer[:emailAddress] = options[:email] || 'unspecified@example.com'
```

shasum replied:
👍

duff commented:
You're right about the implicit defaults for billing address!  

I wonder if there's a way to turn off the `ANTIFRAUD_REJECTED` error for a Brazilian `account_id` in their gateway UI.

Perhaps an option is to have a default shipping_address like you mentioned in [this](https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateways/cyber_source.rb#L252-L262).  So it would use the specified `shipping_address` from the customer, or it would use the hard coded default.  But it would be separate from the `billing_address`.

Hoping I'm not being too nitzy.

shasum commented:
Additional [feedback](https://spreedly.freshdesk.com/helpdesk/tickets/22726) from gateway support. Need to think through changes needed, if any, from Vladimir's observations.

shasum commented:
>I wonder if there's a way to turn off the ANTIFRAUD_REJECTED error for a Brazilian account_id in their gateway UI.

Could be, but no way to check that immediately. Using open test credentials for their sandbox from the API [docs](http://developers.payulatam.com/en/api/sandbox.html) page. There doesn't seem to be a self service portal for the sandbox.
>Perhaps an option is to have a default shipping_address like you mentioned in this. So it would use the specified shipping_address from the customer, or it would use the hard coded default. But it would be separate from the billing_address.

Thinking it might be better to let the error pass through in this case. Allow customers to send in the shipping_address when required. Hard coding an unrelated address feels odd.

duff commented:
@shasum Sounds great!

shasum commented:
@duff end in sight!

Quite a few changes:
- cvv removed from authorization
- billing_address, shipping_address, and email cleaned up
- incorporated all feedback from gateway support
- added query method that helps check status of pending transactions

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+          add_credentials(post, 'SUBMIT_TRANSACTION')
+          add_transaction_type(post, 'REFUND')
+          add_reference(post, authorization)
+        else
```
Same here - I don't know that we need this, do we?

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+        commit('refund', post)
+      end
+
+      # Supported countries have different minimum amounts and
```
I wonder if the currency code factors in.  If they specify the currency in the options, can we lower the amount?

Most gateways have 100 for the `verify` amount.  4000 seems so high!   

I also wonder if for most currencies perhaps 100 might work but for a few currencies we could bump it up to 4000.  Or maybe 4000 all the time is simpler.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+      end
+
+      def verify_credentials
+        response = authorize(4000, "_|VISA|00000000-0000-0000-0000-000000000000", {cvv: '000'})
```
That's a pretty cool idea.  

I wonder if they have an API call that could be used that doesn't take a card.  The main thought I had was would we break if they changed the message text.  Seems like historically, messages are more likely to change than response codes.  But perhaps messages are one of the only ways to distinguish between error states.  Or maybe this message won't change and we'll be fine.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+        response.message.include? "The ID [00000000-0000-0000-0000-000000000000] of credit card token is invalid"
+      end
+
+      def query(authorization, options={})
```
Not sure if we need `query`.  We typically let folks do things like this outside of active merchant since active merchant is primarily for the main X operations like `purchase`, `authorize`, etc.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+      def add_buyer(post, options)
+        buyer = {}
+        buyer[:dniNumber] = options[:buyer_dni_number] if options[:buyer_dni_number]
+        buyer[:cnpj] = options[:buyer_cnpj] if options[:buyer_cnpj]
```
Not sure if we need this, do we?  Same goes for `dniNumber`.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+        post[:transaction][:order][:buyer] = buyer
+      end
+
+      def add_shipping_address(post, options)
```
I know they sent an email about specifying the shipping address more than once.  I do wonder why that's needed!  Is there one that's more important than the other?  Seems weird to be doing it twice, once for the buyer and once for the transaction order.  Would it be bad to only specify it for the transaction order?

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+        tx_tax[:value] = '0'
+        tx_tax[:currency] = options[:currency] || currency(money)
+
+        tx_tax_return_base = {}
```
I saw the email too where they were talking about these TAX fields.  Do transactions fail without all of these `additional_values`?

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+          @options[:api_key],
+          @options[:merchant_id],
+          post[:transaction][:order][:referenceCode],
+          post[:transaction][:order][:additionalValues][:TX_VALUE][:value],
```
The fact that these are part of the signature leads me to think that the TAX values are required.  :(

shasum replied:
It plays tricks on the eyes, but it's the TX_VALUE ( transaction value? ) that's part of signature. Got rid of all other TAX related values.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+
+      def add_payer(post, options)
+        payer = {}
+        payer[:dniNumber] = options[:payer_dni_number] if options[:payer_dni_number]
```
Necessary?

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+      def add_payer(post, options)
+        payer = {}
+        payer[:dniNumber] = options[:payer_dni_number] if options[:payer_dni_number]
+        payer[:cnpj] = options[:payer_cnpj] if options[:payer_cnpj]
```
I'm not sure what `dniNumber` and `payer_cnpj` represent or how likely customers are to use them.  Do you think it's likely someone might want them?

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+        payer = {}
+        payer[:dniNumber] = options[:payer_dni_number] if options[:payer_dni_number]
+        payer[:cnpj] = options[:payer_cnpj] if options[:payer_cnpj]
+        payer[:emailAddress] = options[:email] || 'unspecified@example.com'
```
So `email` gets specified for transaction->payer and for transaction->order->buyer.    I wonder if both are needed.

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,484 @@
+      end
+
+      def add_browser_data(post, options)
+        post[:transaction][:deviceSessionId] = 'n/a'
```
Do the transactions fail if `deviceSessionId`, `cookie`, and `userAgent` aren't specified.  If so, the 👍 on filling them the with `n/a`.

duff commented:
Nice!  I looked through and made a few comments, asked a few questions for background, etc.

All in all, it's looking pretty great; I'm just pushing a little bit to see if there's anything we can simplify.

shasum commented:
@duff the no frills version!

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,400 @@
+        JSON.parse(body)
+      end
+
+      def success_from(action, response)
```
I wonder if `response["code"] == "SUCCESS"` would work for all of 'em.  Are there cases where the response code is SUCCESS and there isn't a `creditCardToken` or a `transactionResponse`?

shasum replied:
For `auth` and `purchase`, it checks `transactionResponse.state` is Approved. 

iirc from a past error, validating that response has the item that we need (eg: `tokenId`) is an approach I adopted to prevent runtime `nil` errors when we subsequently read that value.

duff replied:
I wonder if... there is no `transactionResponse` info and response["code"] == SUCCESS, did money actually actually change hands?  If so, I'm wondering if we want to capture that as a successful transaction and then handle the nil `transactionResponse` subsequently?

shasum replied:
with `code == SUCCESS`, `transactionResponse.state` can be one of `APPROVED`, `DECLINED` or `PENDING`. Only the first one is a successful transaction, which cannot be determined using `code` alone.

duff replied:
Sounds good @shasum.  Thanks for the clarification!  Different gateways seem to have a varying number of fields that need to be checked for success. 😢

duff commented on lib/active_merchant/billing/gateways/payu_latam.rb:
```diff
@@ -0,0 +1,400 @@
+        end
+      end
+
+      def message_from(action, success, response)
```
Any good ideas how we might refactor this a bit?  Perhaps extracting parts into other methods?

shasum replied:
Yep, could use a better refactor. Will do a separate PR.

duff replied:
Sounds great! 👍

duff commented:
@shasum I made a comment or two and then merged it since those comments were nice to haves (and those could be new PR's if you have ideas as to how to address them).
